Library/Homebrew/cli/named_args.rb

```
@@ -29,13 +42,212 @@ def initialize(*args, parent: Args.new, override_spec: nil, force_bottle: false,
+      # @api public
+      sig do
+        params(
+          only:                         T.nilable(Symbol),
+          separate_types:               T::Boolean,
+          ignore_unavailable:           T::Boolean,
+          allow_formula_cask_conflicts: T::Boolean,
+        ).returns(T.any(Array, Hash))
+      end
+      def convert(only: nil, separate_types: false, ignore_unavailable: false, allow_formula_cask_conflicts: false)
+        raise "Unable to convert named args of type `#{@type}`" if @type.present? && NAMED_ARGS_TYPES.exclude?(@type)
+
+        formulae = []
+        casks = []
+        kegs = []
+        taps = []
+        others = []
+
+        @args_to_convert = downcased_unique_named
+
+        only ||= parent&.only_formula_or_cask
+        type = @type if only.blank?
+        type ||= case @type
+        when :formula_or_cask
+          only
+        when :installed_formula_or_cask
+          :"installed_#{only}"
+        else
+          @type
+        end
+
+        case type
+        when :formula
+          formulae += formulae ignore_unavailable: ignore_unavailable
+        when :installed_formula
+          formulae += formulae installed: true, ignore_unavailable: ignore_unavailable
+        when :cask
+          casks += casks ignore_unavailable: ignore_unavailable
+        when :installed_cask
+          casks += casks installed: true, ignore_unavailable: ignore_unavailable
+        when :formula_or_cask
+          formulae += formulae ignore_unavailable: true
+          casks += casks ignore_unavailable: true
+        when :installed_formula_or_cask
+          formulae += formulae installed: true, ignore_unavailable: true
+          casks += casks installed: true, ignore_unavailable: true
+        when :tap
+          taps += taps ignore_unavailable: ignore_unavailable
+        when :installed_tap
+          taps += taps installed: true, ignore_unavailable: ignore_unavailable
+        else
+          others += @args
```

Should this `raise` if this value is invalid?

---

Under my current thinking: no.

We do want to allow named arguments that don't fall into any of the categories. I also current used `named_args :command` and `named_args :diagnostic_check` for some commands. These symbols will cause the args to be treated simply as text, but will (in the future) allow completions to autofill commands or diagnostic checks.

---

Although, on second thought, `args.named` should be sufficient in these cases and we probably shouldn't be calling `args.named.convert`, so maybe that's a good idea.

---

> These symbols will cause the args to be treated simply as text, but will (in the future) allow completions to autofill commands or diagnostic checks.

I feel like we can add this in future when we want but "treating symbols as text" seems overly permissive for now.